simplekv/net/azurestore.py

```diff
@@ -128,3 +121,62 @@ def pickle_azure_store(store):
                              store.create_if_missing)
 
 copyreg.pickle(AzureBlockBlobStore, pickle_azure_store)
+
+
+class IOInterface(io.BufferedIOBase):
+    """
+    Class which provides a file-like interface to selectively read from a blob in the blob store.
+    """
+    def __init__(self, block_blob_service, container_name, key):
+        super(IOInterface, self).__init__()
+        self.block_blob_service = block_blob_service
+        self.container_name = container_name
+        self.key = key
+
+        blob = self.block_blob_service.get_blob_properties(container_name, key)
+        self.size = blob.properties.content_length
+        self.pos = 0
+
+    def tell(self):
+        """Returns he current offset as int. Always >= 0."""
+        return self.pos
+
+    def read(self, size=-1):
+        """Returns 'size' amount of bytes or less if there is no more data.
+        If no size is given all data is returned. size can be >= 0."""
+        with map_azure_exceptions(key=self.key):
+            if size < 0:
+                size = self.size - self.pos
+
+            end = min(self.pos + size - 1, self.size)
+            b = self.block_blob_service.get_blob_to_bytes(
+                    self.container_name,
+                    self.key,
+                    start_range=self.pos,
+                    end_range=end)  # end_range is inclusive
+            self.pos += len(b.content)
+            return b.content
+
+    def seek(self, offset, whence=0):
+        """Move to a new offset either relative or absolute. whence=0 is
+        absolute, whence=1 is relative, whence=2 is relative to the end.
+
+        Any relative or absolute seek operation which would result in a
+        negative position is undefined and that case can be ignored
+        in the implementation.
+
+        Any seek operation which moves the position after the stream
+        should succeed. tell() should report that position and read()
+        should return an empty bytes object."""
```

Comment:
Where does this specification come from? Is there a precedent for this behaviour of `seek()`?

Reply:
https://docs.python.org/3/tutorial/inputoutput.html

To change the file object’s position, use f.seek(offset, from_what). The position is computed from adding offset to a reference point; the reference point is selected by the from_what argument. A from_what value of 0 measures from the beginning of the file, 1 uses the current file position, and 2 uses the end of the file as the reference point. from_what can be omitted and defaults to 0, using the beginning of the file as the reference point.

Reply:
I'm asking because I can not find any description of this behaviour ("Any seek operation which moves the position after the stream should succeed."), and the Python docs only describe it as "undefined" for text files, no word about binary. "In text files [...] the only valid offset values are those returned from the f.tell(), or zero. Any other offset value produces undefined behaviour."

Reply:
This specification is not in the documentation, but that behaviour can be derived from the way python (at least 3.6) behaves on regular files opened in binary mode. Not the best argument, I know, but not bad either.
Actually, we were not behaving like that, so I added code+tests for that.